sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample15_MockingClientTypes.md

```diff
@@ -0,0 +1,619 @@
+    message: message,
+    receiver: mockReceiver.Object,
+    cancellationToken: CancellationToken.None);
+
+Assert.DoesNotThrowAsync(async () => await MessageHandler(processArgs));
+
+string fullyQualifiedNamespace = "full-qualified-namespace";
+string entityPath = "entity-path";
+ServiceBusErrorSource errorSource = new();
+
+ProcessErrorEventArgs errorArgs = new(
+    exception: new Exception("sample exception"),
+    errorSource: errorSource,
+    fullyQualifiedNamespace: fullyQualifiedNamespace,
+    entityPath: entityPath,
+    cancellationToken: source.Token);
+
+Assert.DoesNotThrowAsync(async () => await ErrorHandler(errorArgs));
+```
+
+For the `ServiceBusSessionProcessor`:
+
+```C# Snippet:ServiceBus_TestSessionProcessorHandlers
+Mock<ServiceBusSessionReceiver> mockSessionReceiver = new();
+
+mockSessionReceiver
+    .Setup(receiver => receiver.CompleteMessageAsync(
+        It.IsAny<ServiceBusReceivedMessage>(),
+        It.IsAny<CancellationToken>()))
+    .Returns(Task.CompletedTask);
+
+async Task MessageHandler(ProcessSessionMessageEventArgs args)
+{
+    var body = args.Message.Body.ToString();
+
+    // we can evaluate application logic and use that to determine how to settle the message.
+    await args.CompleteMessageAsync(args.Message);
+
+    // we can also set arbitrary session state using this receiver
+    // the state is specific to the session, and not any particular message
+    await args.SetSessionStateAsync(new BinaryData("some state"));
+}
+
+Task ErrorHandler(ProcessErrorEventArgs args)
+{
+    // the error source tells me at what point in the processing an error occurred
+    Console.WriteLine(args.ErrorSource);
+    // the fully qualified namespace is available
+    Console.WriteLine(args.FullyQualifiedNamespace);
+    // as well as the entity path
+    Console.WriteLine(args.EntityPath);
+    Console.WriteLine(args.Exception.ToString());
+    return Task.CompletedTask;
+}
+
+ServiceBusReceivedMessage message = ServiceBusModelFactory.ServiceBusReceivedMessage(
+        body: new BinaryData("message"),
+        messageId: "messageId",
+        partitionKey: "hellokey",
+        correlationId: "correlationId",
+        contentType: "contentType",
+        replyTo: "replyTo"
+        //...
+        );
+
+ProcessSessionMessageEventArgs processArgs = new(
+    message: message,
+    receiver: mockSessionReceiver.Object,
+    cancellationToken: CancellationToken.None);
+
+Assert.DoesNotThrowAsync(async () => await MessageHandler(processArgs));
+
+string fullyQualifiedNamespace = "full-qualified-namespace";
+string entityPath = "entity-path";
+ServiceBusErrorSource errorSource = new();
+
+ProcessErrorEventArgs errorArgs = new(
+    exception: new Exception("sample exception"),
+    errorSource: errorSource,
+    fullyQualifiedNamespace: fullyQualifiedNamespace,
+    entityPath: entityPath,
+    cancellationToken: CancellationToken.None);
+
+Assert.DoesNotThrowAsync(async () => await ErrorHandler(errorArgs));
```

Comment:
It'd be great if we added the super duper "how do I pretend that my processor is running all by iteself" snippet, like [this one](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/Sample08_MockingClientTypes.md#simulating-the-eventprocessorclient-running).    I have no idea why, but it seems to come up frequently despite guidance to just test the handlers.

Reply:
I would think that scenario would be more common than wanting to test the handlers. From my perspective at least, all these mocking examples help customers unit test _their_ code. They don't necessarily want to test the Service Bus library, but they do need to be able to mock out all of the calls to it.

Reply:
I agree that *should* be the goal - but feedback is about 50/50 for Event Hubs.   Convincing people to just test their code and trust us to call the handlers as appropriate has proven to be more difficult than one would think.    

I certainly agree that the current "test your handlers by calling them directly" is what we want to highlight, but the alternate approach has saved me many words.